pkg/provider/cloud/vmwareclouddirector/helper.go

```diff
@@ -23,12 +23,18 @@ import (
 )
 
 func deleteVApp(vdc *govcd.Vdc, vapp *govcd.VApp) error {
+	waitForUndeploy := true
 	task, err := vapp.Undeploy()
 	if err != nil {
-		return fmt.Errorf("failed to undeploy vApp before deletion: %w", err)
+		// Undeploy failed, it's still safe to delete vApp directly since it will take care of all the cleanup.
+		// Most common reason for failure is that the vApp is not in "running" state.
+		waitForUndeploy = false
 	}
-	if err = task.WaitTaskCompletion(); err != nil {
-		return fmt.Errorf("error waiting for vApp undeploy to complete: %w", err)
+
+	if waitForUndeploy {
+		if err = task.WaitTaskCompletion(); err != nil {
+			return fmt.Errorf("error waiting for vApp undeploy to complete: %w", err)
+		}
```

Comment:
How about 

```go
func deleteVApp(vdc *govcd.Vdc, vapp *govcd.VApp) error {
	// Undeploy can fail, but it's still safe to delete vApp later directly since it will take care of all the cleanup.
	// Most common reason for failure is that the vApp is not in "running" state.
	task, err := vapp.Undeploy()
	if err == nil {
		if err = task.WaitTaskCompletion(); err != nil {
			return fmt.Errorf("error waiting for vApp undeploy to complete: %w", err)
		}
	}

	// ...
```

Reply:
Updated